pub/lib/dependabot/pub/requirement.rb

```diff
@@ -0,0 +1,122 @@
+# frozen_string_literal: true
+
+# For details on pub version constraints see:
+# https://github.com/dart-lang/pub_semver
+
+###################################################################
+
+require "dependabot/utils"
+require "dependabot/pub/version"
+
+module Dependabot
+  module Pub
+    class Requirement < Gem::Requirement
+      quoted = OPS.keys.map { |k| Regexp.quote(k) }.join("|")
+      version_pattern = Pub::Version::VERSION_PATTERN
+
+      PATTERN_RAW = "\\s*(#{quoted})?\\s*(#{version_pattern})\\s*"
+      PATTERN = /\A#{PATTERN_RAW}\z/.freeze
+
+      # Use Pub::Version rather than Gem::Version to ensure that
+      # pre-release versions aren't transformed.
+      def self.parse(obj)
```

Comment:
It's unfortunate that we need to do this at all, I quickly looked if there's a way to avoid doing this for pub, since most of the logic is handled by pub itself, but I'm worried that lot's of things will break in dependabot's current form. Eventually though I think we should be able to get rid of this, once we have a higher level API that we can implement against.